File: sources/scalac/transformer/UnCurryPhase.java
/*     ____ ____  ____ ____  ______                                     *\
**    / __// __ \/ __// __ \/ ____/    SOcos COmpiles Scala             **
**  __\_ \/ /_/ / /__/ /_/ /\_ \       (c) 2002, LAMP/EPFL              **
** /_____/\____/\___/\____/____/                                        **
**                                                                      **
** $Id$
\*                                                                      */

package scalac.transformer;

import scalac.*;
import scalac.parser.*;
import scalac.symtab.*;
import scalac.typechecker.Infer;
import scalac.checkers.*;

public class UnCurryPhase extends PhaseDescriptor implements Modifiers {

    private Global global;

    public void initialize(Global global, int id) {
        super.initialize(global, id);
	this.global = global;
    }

    public String name () {
        return "uncurry";
    }

    public String description () {
        return "uncurry function types and applications";
    }

    public String taskDescription() {
        return "uncurried";
    }

    public Phase createPhase(Global global) {
        return new UnCurry(global, this);
    }

    /** - return symbol's transformed type,
     *  - if symbol is a def parameter with transformed type T, return () => T
     */
    public Type transformInfo(Symbol sym, Type tp0) {
	Type tp1 = uncurry(tp0);
	if (sym.isDefParameter()) return global.definitions.functionType(Type.EMPTY_ARRAY, tp1);
	else return tp1;
    }

    /** - (ps_1)...(ps_n)T ==> (ps_1,...,ps_n)T
     */
    Type uncurry(Type tp) {
	switch (tp) {
	case MethodType(Symbol[] params, Type tp1):
	    Type newtp1 = uncurry(tp1);
	    switch (newtp1) {
	    case MethodType(Symbol[] params1, Type tp2):
		Symbol[] newparams = new Symbol[params.length + params1.length];
		System.arraycopy(params, 0, newparams, 0, params.length);
		System.arraycopy(params1, 0, newparams, params.length, params1.length);
		return Type.MethodType(newparams, tp2);
	    default:
		if (newtp1 == tp1) return tp;
		else return Type.MethodType(params, newtp1);
	    }
	case PolyType(Symbol[] tparams, Type tp1):
	    switch (tp1) {
	    case MethodType(_, _):
		Type newtp1 = uncurry(tp1);
		if (newtp1 == tp1) return tp;
		else return Type.PolyType(tparams, newtp1);
	    default:
		Type newtp1 = Type.MethodType(Symbol.EMPTY_ARRAY, tp1);
		if (tparams.length == 0) return newtp1;
		else return Type.PolyType(tparams, newtp1);
	    }
	case OverloadedType(_, _):
	    return new Type.Map() {
		public Type apply(Type t) { return uncurry(t); }
	    }.map(tp);
	default:
	    return tp;
	}
    }

    public Checker[] postCheckers(Global global) {
        return new Checker[] {
            new CheckSymbols(global),
            new CheckTypes(global),
            new CheckOwners(global),
	    new CheckNames(global)
        };
    }
}

File: sources/scala/tools/scalac/transformer/UnCurryPhase.scala
/*     ____ ____  ____ ____  ______                                     *\
**    / __// __ \/ __// __ \/ ____/    SOcos COmpiles Scala             **
**  __\_ \/ /_/ / /__/ /_/ /\_ \       (c) 2002, LAMP/EPFL              **
** /_____/\____/\___/\____/____/                                        **
\*                                                                      */

// $Id: UnCurryPhase.scala

import scalac.{Global => scalac_Global, _};
import scalac.symtab._;
import scalac.checkers._;

package scala.tools.scalac.transformer {

import scalac.util.NewArray;

class UnCurryPhase(global: scalac_Global, descriptor: PhaseDescriptor) extends Phase(global, descriptor) {

  import Modifiers._;

  /** Applies this phase to the given compilation units. */
  override def apply(units: Array[CompilationUnit]) = {
    var i = 0; while (i < units.length) {
      new UnCurry(global, this).apply(units(i));
      i = i + 1
    }
  }

  /** - return symbol's transformed type,
   *  - if symbol is a def parameter with transformed type T, return () => T
   */
  override def transformInfo(sym: Symbol, tp0: Type): Type = {
    val tp1 = uncurry(tp0);
    if (sym.isDefParameter()) global.definitions.FUNCTION_TYPE(Type.EMPTY_ARRAY, tp1);
    else tp1
  }

  /** - (ps_1)...(ps_n)T ==> (ps_1,...,ps_n)T
   */
  def uncurry(tp: Type): Type = tp match {
    case Type$MethodType(params, tp1) =>
      uncurry(tp1) match {
	case Type$MethodType(params1, tp2) =>
	  val newparams = new Array[Symbol](params.length + params1.length);
	  System.arraycopy(params, 0, newparams, 0, params.length);
	  System.arraycopy(params1, 0, newparams, params.length, params1.length);
	  new Type.MethodType(newparams, tp2);
	case newtp1 =>
	  if (newtp1 == tp1) tp;
	  else new Type.MethodType(params, newtp1);
      }
    case Type$PolyType(tparams, tp1) =>
      uncurry(tp1) match {
        case newtp1 @ Type$MethodType(_, _) =>
	  if (newtp1 == tp1) tp
	  else new Type.PolyType(tparams, newtp1)
	case newtp1 =>
	  val newtp2 = Type.MethodType(Symbol.EMPTY_ARRAY, newtp1);
	  if (tparams.length == 0) newtp2;
	  else Type.PolyType(tparams, newtp2)
      }
    case Type$OverloadedType(_, _) =>
      new Type$Map() {
	override def apply(t: Type) = uncurry(t)
      }.map(tp);
    case Type$ConstantType(base, _) =>
      base
    case Type$CompoundType(parents, scope) =>
      val symbol = tp.symbol();
      if (!symbol.isClass() || symbol.isCompoundSym()) tp
      else {
        val clone = new Scope();
        val it = scope.iterator(true);
        while (it.hasNext()) {
          val member = it.next();
          if (!isUnaccessedConstant(member) &&
              (!member.isCaseFactory() || member.isModule()))
            clone.enterOrOverload(member);
        }
        Type.compoundType(parents, clone, symbol)
      }
    case _ =>
      tp
  }

  def isUnaccessedConstant(symbol: Symbol): boolean =
    symbol.isTerm() &&
    (symbol.flags & ACCESSED) == 0 &&
    (symbol.getType() match {
      case Type$PolyType(params, Type$ConstantType(_, _)) =>
        params.length == 0
      case Type$ConstantType(_, _) =>
        true
      case _ =>
        false
    });

  override def postCheckers(global: scalac_Global): Array[Checker] =
    NewArray.Checker(new CheckSymbols(global), new CheckTypes(global));
}
}
